examples/cis2-wccd-upgrade/src/lib.rs

```diff
@@ -0,0 +1,2030 @@
+    state_builder: &mut StateBuilder<S>,
+    logger: &mut impl HasLogger,
+) -> InitResult<StateImplementation<S>> {
+    // Get the instantiater of this contract instance.
+    let invoker = Address::Account(ctx.init_origin());
+    // Construct the initial contract state.
+    let state = StateImplementation::new(invoker, state_builder);
+    // Log event for the newly minted token.
+    logger.log(&Cis2Event::Mint(MintEvent {
+        token_id: TOKEN_ID_WCCD,
+        amount:   ContractTokenAmount::from(0u64),
+        owner:    invoker,
+    }))?;
+
+    // Log event for where to find metadata for the token
+    logger.log(&Cis2Event::TokenMetadata::<_, ContractTokenAmount>(TokenMetadataEvent {
+        token_id:     TOKEN_ID_WCCD,
+        metadata_url: MetadataUrl {
+            url:  String::from(TOKEN_METADATA_URL),
+            hash: None,
+        },
+    }))?;
+
+    Ok(state)
+}
+
+/// Initialzes the implementation of the w_ccd contract with the proxy and the
+/// state addresses.
+#[receive(
+    contract = "CIS2-wCCD",
+    name = "initialize",
+    parameter = "InitializeImplementationParams",
+    mutable
+)]
+fn contract_initialize<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation<S>, StateApiType = S>,
+) -> ContractResult<()> {
+    ensure!(!host.state().initialized, concordium_cis2::Cis2Error::Custom(AlreadyInitialized));
+    // Set proxy and storage addresses.
+    let params: InitializeImplementationParams = ctx.parameter_cursor().get()?;
+    host.state_mut().proxy_address = params.proxy_address;
+    host.state_mut().state_address = params.state_address;
+    // Set the w_ccd_implementation contract as initialized.
+    host.state_mut().initialized = true;
+    Ok(())
+}
+
+/// Wrap an amount of CCD into wCCD tokens and transfer the tokens if the sender
+/// is not the receiver.
+#[receive(
+    contract = "CIS2-wCCD",
+    name = "wrap",
+    parameter = "WrapParams",
+    enable_logger,
+    mutable,
+    payable
+)]
+fn contract_wrap<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation<S>, StateApiType = S>,
+    amount: Amount,
+    logger: &mut impl HasLogger,
+) -> ContractResult<()> {
+    let slot_time = ctx.metadata().slot_time();
+
+    match host.state().state_address {
+        None => bail!(concordium_cis2::Cis2Error::Custom(OnlyState)),
+
+        Some(state_address) => {
+            let unpause_time = host
+                .invoke_contract_raw(
+                    &state_address,
+                    Parameter(&[]),
+                    EntrypointName::new_unchecked("get_unpause_time"),
+                    Amount::zero(),
+                )?
+                .1;
+
+            let unpause_time = if let Some(mut unpause_time) = unpause_time {
+                unpause_time.get()?
+            } else {
+                return Err(concordium_cis2::Cis2Error::Custom(StateInvokeError));
```

Comment:
I recommend commenting a bit on what the logic here is (e.g., that you are expecting a response since the state contract must be a V1 contract).